teps/0146-parameters-in-script.md

```diff
@@ -0,0 +1,542 @@
+
+#### How will we know that this has succeeded?
+We will now this has succeeded:
+  1. When there is a method of securely and reliably uses `Parameters` in `script`s 
+  2. When the current method with less security and reliability is deprecated/no-longer-allowed and using `Parameters` in `script`s is secure by default
+
+### Non-Goals
+Directly disallowing the use of `Param` values in `script`s without providing a feasible alternative.
+
+### Use Cases
+As described in the [original proposal][pipeline-issue-781], the use case for `Script`
+in `Tasks` is:
+
+> Users need an easy cruft-free way to express a multi-statement `Script` in the
+> body of a `Task` `Step`, without having to understand advanced topics about
+> `Containers`, like what an `entrypoint` or a `command` is.
+>
+> Users should only need to have passing familiarity with a shell environment
+> to be successful using Tekton.
+
+Problems with variable subsitution was noted in [original proposal][pipeline-issue-781]:
+
+> How this syntax works with input substitutions, e.g., ``${inputs.params.flags}`` might
+> also lead to confusion among users, which we should design for.
+
+Users need a way to safely and reliably substitute `Parameters` in `Scripts` in `Steps`
+of their `Tasks`. This would make it easier to use Tekton without learning advanced topics,
+while still using it reliably and safely.
+
+### Requirements
+
+<!--
+Describe constraints on the solution that must be met, such as:
+- which performance characteristics that must be met?
+- which specific edge cases that must be handled?
+- which user scenarios that will be affected and must be accommodated?
+-->
+
+1. **[Secure by default in v1][secure-by-default]**: Out of the box
+   *Tekton Pipelines* v1 should disallow common anti-patterns that
+   can render a default install insecure.
+1. **Easy to use**: Whatever replaces the direct use of `$(params.X)` in
+   `script` should be unsurprising and memorable to ease friction in user
+   migration.
+1. **Automatable**: A script or program should be able to convert `Tasks`
+   to the new approach. This might be impossible for all cases (e.g. if an
+   unrecognized shebang has been used in the script) but a best-effort
+   implementation should be provided.
+1. **Backwards-compatible in v1**. If a user already has `Tasks`
+   that use `$(params.X)` syntax in `script` then we should continue supporting
+   that usage in v1 resources. Ideally, Tekton Pipelines would warn the user
+   about it somehow.
+1. **(To be discussed) Provide an escape hatch in v1**. If a cluster
+   operator or admin decides that they want to continue allowing
+   `Parameter` variables to be directly injectable to `script` fields of
+   `Steps` there should be some way for them to configure that.
+
+## Proposal - Implicitly Project `Parameters` as Environment Variables in `Steps` (Language-Agnostic)
+We have projected `Parameters` as environment variables in the `git-clone` Task in the
```

Comment:
I'm not sure how to prevent attack vector injection by using Environment Variables. If there's an attack happening to inject malicious content to that param, that param will also be passed to the env var and there's no check in this process?

Reply:
I have added new sections w/ examples related to the security improvements and reliability improvements uses env vars has vs direct interpolation.  

The gist is that program environment variables are designed in such a way to have additional security around code injection:
- https://mywiki.wooledge.org/BashProgramming/05#Environment_variables ? https://www.redhat.com/en/blog/bash-specially-crafted-environment-variables-code-injection-attack

Reply:
An example from the new section, if you have something like:

```
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  name: echo-message-pipelinerun
spec:
  pipelineRef:
    name: echo-message-pipeline
  params:
    - name: message
      value: "$(echo $(cat /var/run/secrets/kubernetes.io/serviceaccount/token))"
```

```
apiVersion: tekton.dev/v1beta1
kind: Task
metadata:
  name: echo-message-task
spec:
  params:
    - name: message
      type: string
      description: The message to echo
      default: "Hello, World!"
  steps:
    - name: echo
      image: ubuntu
      script: |
        #!/bin/bash
        echo "$(params.message)"
```

If you run this with tekton as it currently is, you will get an output like:
```
$ <the-token-at-/var/run/secrets/kubernetes.io/serviceaccount/token>
```

BUT if you run this and instead of doing direct string interpolation you use an env var, you won't run injected code.  For example, using my branch where I have implemented this solution, this is to log output of the below:

```
kind: Task
metadata:
  name: echo-message-task
spec:
  params:
    - name: message
      type: string
      description: The message to echo
      default: "Hello, World!"
  steps:
    - name: echo
      image: ubuntu
      script: |
        #!/bin/bash
        echo "$(params.message)"
        echo "$$(params.message.env)"
 ```
 
 ```
 $ aprindle@aprindle-ssd ~/repro-params-scripts/string kubectl logs echo-message-pipelinerun-echo-message-pod 
Defaulted container "step-echo" out of: step-echo, prepare (init), place-scripts (init)
$ <the-token-at-/var/run/secrets/kubernetes.io/serviceaccount/token>
$(echo $(cat /var/run/secrets/kubernetes.io/serviceaccount/token))
 ```

Reply:
Oh thanks a lot for the detailed explanations! Learned a lot 😄